ReusableBlockConvertButton: use React hooks.

ZebulanStanphill opened the pull request:
## Description
This PR refactors ReusableBlockConvertButton to use the `useSelect` and `useDispatch` hooks, rather than the HOC equivalents. It also fixes some comment typos and removes an unnecessary Lodash dependency.

Side note: I'm working on a PR to [move the "Convert to Regular Block" button to the block toolbar](https://github.com/WordPress/gutenberg/issues/8403). This PR is intended to polish the code a bit before I make that change.

## How has this been tested?
I've tested to make sure the "Add to Reusable blocks" and "Convert to Reusable Block" options appear when they should and work as intended.

## Types of changes
Only refactoring + code quality changes. Everything should act the same as `master`.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

ZebulanStanphill committed:
ReusableBlockConvertButton: use React hooks.

youknowriad commented on packages/editor/src/components/reusable-blocks-buttons/reusable-block-convert-button.js:
```diff
@@ -1,24 +1,61 @@
+				isReusable ||
+				// Hide 'Add to Reusable blocks' when reusable blocks are disabled.
+				( canInsertBlockType( 'core/block' ) &&
+					blocks?.every(
```
This is interesting while `every(undefined, something)` returns true, `undefined?.every( something )` returns undefined (aka false) which changes the meaning.

It's one of the reasons I still prefer lodash defaults over having to think about these subtile usecases. I find `something?.something` to be very problematic in a lot of usecases and something to use but not to overuse. It can quickly introduce hidden bugs everywhere.

ZebulanStanphill replied:
Lodash's `every(undefined, something)` returning `true` is a surprise to me. That's not at all what I would expect.

In this case, though, it seems like it isn't even possible for `blocks` to be undefined; so the optional chaining shouldn't even be necessary, right?

youknowriad replied:
Mathematically speaking Every(undefined) = true is the correct thing.

> In this case, though, it seems like it isn't even possible for blocks to be undefined; so the optional chaining shouldn't even be necessary, right? 

IMO, we  should assume things like that. We don't really know at this point what we have in store even if it seems technically impossible. The only way to ensure that is to have typing everywhere. 

Maybe:

 - When mounting/unmounting blocks, there's a delay where it could happen,
 - Maybe there's some async behavior (there isn't) causing this.

What I'm trying to say is that the component should be written consistently regardless of how the rest of the system works.

ZebulanStanphill replied:
In my opinion, running `every` on `undefined` isn't something one would assume to be possible by default. That sounds like just as much of a potential gotcha as the optional chaining resulting in `false` by default.

In this case, I think the current change in behavior is probably fine... if `blocks` is empty, there's nothing to convert to a reusable block at the moment, so we wouldn't want to show the option, right? In fact, that's what I had assumed the logic was already doing before my changes.

Funny enough, while discussing this, I've just found a bug in the code. We're checking for `isReusable || ...` when it should be `_isReusable || ...`. Additionally, instead of using optional chaining, we can just use nullish coalescing to default `blocks` to `[]` when it is first defined.

I'm going to go implement both of those changes real quick.

youknowriad reviewed (approved):
Approving but I think we should restore lodash here personally.

ZebulanStanphill committed:
Fix/simplify _isVisible initialization.

ZebulanStanphill committed:
Add types JSDoc to component.

ZebulanStanphill commented:
Alright, I've fixed the `_isVisible` logic bugs/gotchas and while I was at it, I added a JSDoc to the component for some slightly stronger typing (and to provide a description of what the component does). I'm going to go ahead and merge when the tests pass.

ZebulanStanphill commented:
I'd like a final code review prior to merge. Since the bugs this PR fixes are rather significant (in my opinion), I've added the "Backport to WP Core" label. Feel free to remove it if you disagree.

ZebulanStanphill commented:
Whoops, posted that last comment on the wrong PR! Sorry...
